simple-loadflow/src/main/java/com/powsybl/loadflow/simple/ac/AcloadFlowEngine.java

```diff
@@ -0,0 +1,117 @@
+
+    private final MatrixFactory matrixFactory;
+
+    private final AcLoadFlowObserver observer;
+
+    public AcloadFlowEngine(Network network, LoadFlowParameters parameters, MatrixFactory matrixFactory, AcLoadFlowObserver observer) {
+        this.network = Objects.requireNonNull(network);
+        this.parameters = Objects.requireNonNull(parameters);
+        this.matrixFactory = Objects.requireNonNull(matrixFactory);
+        this.observer = Objects.requireNonNull(observer);
+    }
+
+    private boolean runMacroActions(List<MacroAction> macroActions, MacroIterationContext macroIterationContext) {
+        boolean cont = false;
+
+        for (MacroAction macroAction : macroActions) {
+            observer.beforeMacroActionRun(macroIterationContext.getMacroIteration(), macroAction.getName());
+
+            cont |= macroAction.run(macroIterationContext);
+
+            observer.afterMacroActionRun(macroIterationContext.getMacroIteration(), macroAction.getName());
+        }
+
+        observer.endMacroIteration(macroIterationContext.getMacroIteration());
+
+        return cont;
+    }
+
+    public AcLoadFlowResult run() {
+        Stopwatch stopwatch = Stopwatch.createStarted();
+
+        SimpleAcLoadFlowParameters parametersExt = parameters.getExtension(SimpleAcLoadFlowParameters.class);
+        if (parametersExt == null) {
+            parametersExt = new SimpleAcLoadFlowParameters();
+        }
+
+        List<MacroAction> macroActions = new ArrayList<>();
+        if (parametersExt.isDistributedSlack()) {
+            macroActions.add(new DistributedSlackAction());
+        }
+
+        List<NetworkContext> networkContexts = NetworkContext.of(network, parametersExt.getSlackBusSelectionMode());
+
+        // only process main (largest) connected component
+        NetworkContext networkContext = networkContexts.get(0);
+
+        NewtonRaphsonParameters newtonRaphsonParameters = new NewtonRaphsonParameters()
+                .setVoltageInitMode(parameters.getVoltageInitMode());
+
+        observer.beforeEquationSystemCreation();
+
+        EquationContext equationContext = new EquationContext();
+        EquationSystem equationSystem = AcEquationSystem.create(networkContext, equationContext);
+
+        observer.afterEquationSystemCreation();
+
+        int macroIteration = 0;
+        NewtonRaphsonResult newtonRaphsonResult = null;
+        do {
+            observer.beginMacroIteration(macroIteration);
+
+            int newtowRaphsonIteration = newtonRaphsonResult != null ? newtonRaphsonResult.getIteration() : 0;
+            newtonRaphsonResult = new NewtonRaphson(networkContext, matrixFactory, observer, equationContext,
+                                                    equationSystem, newtonRaphsonParameters, newtowRaphsonIteration)
+                    .run();
+        } while (runMacroActions(macroActions, new MacroIterationContext(macroIteration, networkContext, newtonRaphsonResult)));
```

Comment:
Actually I think macro actions could be nested :
for example we don't want voltage control and frequency regulation to act at the same time after each Newton Raphson, but only activate the second one when the first one has "stabilized".
So with 2 "macro actions" defined, we would have the same behaviour as with 2 nested while loop.

Reply:
Thanks, I changed the design to support nested macro actions.